libvast/src/format/zeek.cpp

```diff
@@ -554,64 +456,159 @@ writer::writer(path dir) {
 }
 
 writer::~writer() {
-  if (streams_.empty())
-    return;
-  std::ostringstream ss;
-  ss << "#close" << separator << time_factory{} << '\n';
-  auto footer = ss.str();
-  for (auto& pair : streams_)
-    if (pair.second)
-      *pair.second << footer;
+  // nop
 }
 
-expected<void> writer::write(const event& e) {
-  if (!caf::holds_alternative<record_type>(e.type()))
-    return make_error(ec::format_error, "cannot process non-record events");
-  std::ostream* os = nullptr;
+namespace {
+
+template <class Iterator>
+class zeek_printer : public printer<zeek_printer<Iterator>> {
+public:
+  using attribute = view<data>;
+
+  bool operator()(Iterator& out, view<caf::none_t>) const {
+    *out++ = unset_field;
+    return true;
+  }
+
+  template <class T>
+  bool operator()(Iterator& out, const T& x) const {
+    if constexpr (detail::is_any_v<T, view<vector>, view<set>>) {
+      if (x.empty()) {
+        for (auto c : std::string_view(empty_field))
+          *out++ = c;
+        return true;
+      }
+      auto i = x.begin();
+      if (!print(out, *i))
+        return false;
+      for (++i; i != x.end(); ++i) {
+        *out++ = set_separator;
+        if (!print(out, *i))
+          return false;
+      }
+      return true;
+    } else if constexpr (std::is_same_v<T, view<map>>) {
+      VAST_ERROR(this, "cannot print maps in Zeek format");
+      return false;
+    } else {
+      make_printer<T> p;
+      return p.print(out, x);
+    }
+  }
+
+  bool operator()(Iterator& out, const view<real>& x) const {
+    return real_printer_.print(out, x);
+  }
+
+  bool operator()(Iterator& out, const view<time>& x) const {
+    double d;
+    convert(x.time_since_epoch(), d);
+    return real_printer_.print(out, d);
+  }
+
+  bool operator()(Iterator& out, const view<duration>& x) const {
+    double d;
+    convert(x, d);
+    return real_printer_.print(out, d);
+  }
+
+  bool operator()(Iterator& out, const view<std::string>& str) const {
+    for (auto c : str)
+      if (!std::isprint(c) || c == separator || c == set_separator) {
+        auto hex = detail::byte_to_hex(c);
+        *out++ = '\\';
+        *out++ = 'x';
+        *out++ = hex.first;
+        *out++ = hex.second;
+      } else {
+        *out++ = c;
+      }
+    return true;
+  }
+
+  bool operator()(Iterator& out, const view<port>& p) const {
+    return (*this)(out, count{p.number()});
+  }
+
+  bool print(Iterator& out, const attribute& d) const {
+    auto f = [&](const auto& x) { return (*this)(out, x); };
+    return caf::visit(f, d);
+  }
+
+private:
+  real_printer<real, 6, 6> real_printer_;
```

Comment:
This suggests that we're dealing with a stateful parser. I'd say either create a type alias at class scope or even a static variable:

```cpp
static constexpr inline auto zeek_real = real_printer<real, 6, 6>{}:
```